### ISSUE TITLE
servicePath is not fully supported by Registry contract

### ISSUE THREAD

**Issue by astroseger** [singnet/platform-contracts#60]
It seems, that in the current implementation we cannot have two services with the same name but different path:
* path1/service
* path2/service
We can register it, but getServiceRegistrationByName will return only one service.

**Comment by ksridharbabuus** [comment]
My suggestion is to refer the services with unique Id may be a guid rather
than with a service name.
BTW why do we need to have two different end points as daemon can speak to
LB which in turn can interface with the services based LB algorithm.
We need to look at how are we going to handle in case of multiple region
deployments.. in this scenario we need to look at some thing like a traffic
manager across the regions.
Also in case of multiple versions of services might have different
instances running in parallel.

**Comment by vsbogd** [comment]
Hi @ksridharbabuus, I am not sure I understand your question properly. Are you asking about service endpoints which are located in service metadata? 

These endpoints are endpoints which are listened by daemon. Having multiple endpoints allow client choosing:
- daemon instance which is located in same region
- daemon instance which is up if some of them are down
So this is basically a way to setup multi-regional service: to have few daemon endpoints in each region.

I would propose using separate issue for further discussion, as this one is about registry paths.

**Comment by tiero** [comment]
> It seems, that in the current implementation we cannot have two services with the same name but different path:
> 
>     * path1/service
> 
>     * path2/service
>       We can register it, but getServiceRegistrationByName will return only one service.

I'm closing it, It seems to not be a bug at all. 

The `getServiceRegistrationByName` returns on purpose a single record, the name it's a combination of `<organization>/<service>`, being already a unique id so to speak.

You can query all the records `listServicesForOrganization` and find the service in the list returned, as it should be, off-chain.

**Comment by astroseger** [comment]
@tiero  Please don't close issues without discussion. 

Sorry, but it is a "bug", maybe not directly in contracts, but in the system as a whole. 

People will solely use getServiceRegistrationByName, and they will rely on it (it is the current state in snet-cli). But it only works in the case when we don't have service with the same name. It means that their code will not work correctly in the case then we have such collision. I think it is the case in snet-cli now.
So there are following solutions.
1. We should well communicate that getServiceRegistrationByName will not work in situation of name collisions, and they should check this collision by their self.
2. We could remove paths, and make them parts of names
3. We could remove getServiceRegistrationByName or make it getServiceRegistrationByNameAndPath

**Comment by arturgontijo** [comment]
@astroseger I think I don't get this, please could you give me (maybe us) an example when the user will want to publish more than one `<org_name><service_name>`?

**Comment by arturgontijo** [comment]
Now I think I understood the issue.
So if the user face the case (2), right now, he will be forced to create another agent address (different `<org_name><service_name>` with the new path? Or he can publish the new path on the same agent address but he can't retrieve this new info (second path) from Registry?

What about to pass a list of paths when publishing/updating the service?

**Comment by vsbogd** [comment]
Actually path is already present in create and update service methods:
```sol
function createServiceRegistration(bytes32 orgName, bytes32 serviceName, bytes32 servicePath, address agentAddress, bytes32[] tags) external {
function updateServiceRegistration(bytes32 orgName, bytes32 serviceName, bytes32 servicePath, address agentAddress) external {
```
As far as I see servicePath is an service attribute. So case (2) is actually impossible: you cannot publish two services in same organization using same name but different path.

For me it is not clear what is the role of the servicePath then?
May be we can just remove it?

**Comment by vsbogd** [comment]
Or move it to metadata?

**Comment by ferrouswheel** [comment]
@astroseger's example is what I almost ran into. My plan was to put services related to human faces all under the `faces` path. And then use service names like `detect`, `identify`, `landmarks` under that path. Such generic service names are likely to be reused under different paths.

If the path component doesn't provide any sort of namespacing, then I would question what purpose it has? It'd be better to just get rid of paths completely then trick developers into thinking they keep things separated.

(A design question is whether paths should be use to group services in this way, and we should give some guidance and examples in singnet/wiki#13)

**Comment by pennachin** [comment]
I agree with @astroseger and @ferrouswheel this is a bug, we should fully support paths.

**Comment by astroseger** [comment]
was done by #70 by removing path from Registry. Now we have a displayname in metadata which could include path.
